tests/components/http/test_forwarded.py

```diff
@@ -0,0 +1,487 @@
+
+
+async def test_x_forwarded_for_with_untrusted_proxy(aiohttp_client):
+    """Test that we get the IP from transport with untrusted proxy."""
+
+    async def handler(request):
+        url = mock_api_client.make_url("/")
+        assert request.host == f"{url.host}:{url.port}"
+        assert request.scheme == "http"
+        assert not request.secure
+        assert request.remote == "127.0.0.1"
+
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+    async_setup_forwarded(app, [ip_network("1.1.1.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get("/", headers={X_FORWARDED_FOR: "255.255.255.255"})
+
+    assert resp.status == 200
+
+
+async def test_x_forwarded_for_with_spoofed_header(aiohttp_client):
+    """Test that we get the IP from the transport with a spoofed header."""
+
+    async def handler(request):
+        url = mock_api_client.make_url("/")
+        assert request.host == f"{url.host}:{url.port}"
+        assert request.scheme == "http"
+        assert not request.secure
+        assert request.remote == "255.255.255.255"
+
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/", headers={X_FORWARDED_FOR: "222.222.222.222, 255.255.255.255"}
+    )
+
+    assert resp.status == 200
+
+
+@pytest.mark.parametrize(
+    "x_forwarded_for",
+    [
+        "This value is invalid",
+        "1.1.1.1, , 1.2.3.4",
+        "1.1.1.1,,1.2.3.4",
+        "1.1.1.1, batman, 1.2.3.4",
+        "192.168.0.0/24",
+        "192.168.0.0/24, 1.1.1.1",
+        ",",
+        "",
+    ],
+)
+async def test_x_forwarded_for_with_malformed_header(
+    x_forwarded_for, aiohttp_client, caplog
+):
+    """Test that we get a HTTP 400 bad request with a malformed header."""
+    app = web.Application()
+    app.router.add_get("/", mock_handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+
+    resp = await mock_api_client.get("/", headers={X_FORWARDED_FOR: x_forwarded_for})
+
+    assert resp.status == 400
+    assert "Invalid IP address in X-Forwarded-For" in caplog.text
+
+
+async def test_x_forwarded_for_with_multiple_headers(aiohttp_client, caplog):
+    """Test that we get a HTTP 400 bad request with multiple headers."""
+    app = web.Application()
+    app.router.add_get("/", mock_handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+
+    resp = await mock_api_client.get(
+        "/",
+        headers=[
+            (X_FORWARDED_FOR, "222.222.222.222"),
+            (X_FORWARDED_FOR, "123.123.123.123"),
+        ],
+    )
+
+    assert resp.status == 400
+    assert "Too many headers for X-Forwarded-For" in caplog.text
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)